plugins/database/mongodb/connection_producer.go

```diff
@@ -187,3 +156,112 @@ func (c *mongoDBConnectionProducer) secretValues() map[string]interface{} {
 		c.Password: "[password]",
 	}
 }
+
+func (c *mongoDBConnectionProducer) getConnectionURL() (connURL string) {
+	connURL = dbutil.QueryHelper(c.ConnectionURL, map[string]string{
+		"username": c.Username,
+		"password": c.Password,
+	})
+	return connURL
+}
+
+func (c *mongoDBConnectionProducer) getWriteConcern() (opts *options.ClientOptions, err error) {
+	if c.WriteConcern == "" {
+		return nil, nil
+	}
+
+	input := c.WriteConcern
+
+	// Try to base64 decode the input. If successful, consider the decoded
+	// value as input.
+	inputBytes, err := base64.StdEncoding.DecodeString(input)
+	if err == nil {
+		input = string(inputBytes)
+	}
+
+	concern := &writeConcern{}
+	err = json.Unmarshal([]byte(input), concern)
+	if err != nil {
+		return nil, errwrap.Wrapf("error unmarshalling write_concern: {{err}}", err)
+	}
+
+	// Translate write concern to mongo options
+	var w writeconcern.Option
+	switch {
+	case concern.W != 0:
+		w = writeconcern.W(concern.W)
+	case concern.WMode != "":
+		w = writeconcern.WTagSet(concern.WMode)
+	default:
+		w = writeconcern.WMajority()
+	}
+
+	var j writeconcern.Option
+	switch {
+	case concern.FSync:
+		j = writeconcern.J(concern.FSync)
+	case concern.J:
+		j = writeconcern.J(concern.J)
+	default:
+		j = writeconcern.J(false)
+	}
+
+	writeConcern := writeconcern.New(
+		w,
+		j,
+		writeconcern.WTimeout(time.Duration(concern.WTimeout)*time.Millisecond))
+
+	opts = options.Client()
+	opts.SetWriteConcern(writeConcern)
+	return opts, nil
+}
+
+func (c *mongoDBConnectionProducer) getAuth() (opts *options.ClientOptions, err error) {
+	if len(c.TLSCAData) == 0 && len(c.TLSCertificateKeyData) == 0 {
+		return nil, nil // NOOP since the user/pass should be in the connection string
+	}
+	return c.getTLSAuth()
+}
+
+func (c *mongoDBConnectionProducer) getUserPassAuth() (opts *options.ClientOptions, err error) {
+	return nil, nil
+}
+
+func (c *mongoDBConnectionProducer) getTLSAuth() (opts *options.ClientOptions, err error) {
+	if len(c.TLSCAData) == 0 && len(c.TLSCertificateKeyData) == 0 {
+		return nil, nil
+	}
+
+	opts = options.Client()
+
+	tlsConfig := &tls.Config{}
+
+	if len(c.TLSCAData) > 0 {
+		if tlsConfig.RootCAs == nil {
```

Comment:
This is always true since we just made `tlsConfig`. Might as well initialize on 237 with the CertPool.

Reply:
Removed the conditional. I didn't add it to the default `tlsConfig` because then it would apply when only using the `TLSCertificateKeyData` rather than using the default in that branch.